### PR TITLE
[ospp] support custom notice template

### PR DESCRIPTION
**Pull request by Eden4701**
## What's changed?
Custom Notice Template Configuration

## Checklist

- [x]  I have read the [Contributing Guide](https://hertzbeat.com/docs/others/contributing/)
- [x]  I have written the necessary doc or comment.
- [x]  I have added the necessary unit tests and all cases have passed.

### CODE REVIEW

File: common/src/main/java/org/dromara/hertzbeat/common/entity/manager/NoticeTemplate.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+    @Schema(title = "Notification Template Entity Primary Key Index ID",
+            description = "通知模板实体主键索引ID",
+            example = "87584674384", accessMode = READ_ONLY)
+    private Long id;
+
+    @Schema(title = "Template name",
+            description = "模板名称",
+            example = "dispatch-1", accessMode = READ_WRITE)
+    @Length(max = 100)
+    @NotNull
```
Comment (gcdd1993) [inline common/src/main/java/org/dromara/hertzbeat/common/entity/manager/NoticeTemplate.java:67]: 针对String类型的非空校验，可以使用`@NotBlank`

File: common/src/main/java/org/dromara/hertzbeat/common/entity/manager/NoticeTemplate.java [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+
+    @Schema(title = "This record creation time (millisecond timestamp)",
+            description = "记录创建时间", accessMode = READ_ONLY)
+    @CreatedDate
+    private LocalDateTime gmtCreate;
+
+    @Schema(title = "Record the latest modification time (timestamp in milliseconds)",
+            description = "记录最新修改时间", accessMode = READ_ONLY)
+    @LastModifiedDate
+    private LocalDateTime gmtUpdate;
```
Comment (gcdd1993) [inline common/src/main/java/org/dromara/hertzbeat/common/entity/manager/NoticeTemplate.java:109]: 是否缺少一个字段，用于标识是否系统预设模版（系统预设模板不允许删除）？

File: manager/src/main/java/org/dromara/hertzbeat/manager/service/impl/NoticeConfigServiceImpl.java [hunk 3]
```diff
@@ -189,7 +344,9 @@ public boolean sendTestMsg(NoticeReceiver noticeReceiver) {
         alert.setFirstAlarmTime(System.currentTimeMillis());
         alert.setLastAlarmTime(System.currentTimeMillis());
         alert.setPriority(CommonConstants.ALERT_PRIORITY_CODE_CRITICAL);
-        return dispatcherAlarm.sendNoticeMsg(noticeReceiver, alert);
+        Byte type=noticeReceiver.getType();
+        NoticeTemplate noticeTemplate=getNoticeTemplatesById(Long.valueOf(type));;
```
Comment (gcdd1993) [inline manager/src/main/java/org/dromara/hertzbeat/manager/service/impl/NoticeConfigServiceImpl.java:348]: 这边逻辑不太对，传参是`type`，但是查询是通过模板ID查询

File: manager/src/main/java/org/dromara/hertzbeat/manager/service/impl/NoticeConfigServiceImpl.java [hunk 4]
```diff
@@ -180,6 +312,29 @@ public NoticeRule getNoticeRulesById(Long ruleId) {
...
+
+    @Override
+    public void deleteNoticeTemplate(Long templateId) {
+        noticeTemplateDao.deleteById(templateId);
+        clearNoticeRulesCache();
+    }
+
+    @Override
+    public NoticeTemplate getNoticeTemplatesById(Long templateId) {
+        return noticeTemplateDao.getReferenceById(templateId);
```
Comment (gcdd1993) [inline manager/src/main/java/org/dromara/hertzbeat/manager/service/impl/NoticeConfigServiceImpl.java:335]: 建议使用`findById`，返回值用`Optional`

File: manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/AbstractAlertNotifyHandlerImpl.java [hunk 5]
```diff
@@ -17,78 +17,97 @@
...
+        model.put("title",  bundle.getString("alerter.notify.title"));
+
         if (alert.getTags() != null) {
             alert.getTags().forEach(context::setVariable);
         }
-        context.setVariable("monitorIdLabel", bundle.getString("alerter.notify.monitorId"));
-        context.setVariable("monitorNameLabel", bundle.getString("alerter.notify.monitorName"));
-        context.setVariable("targetLabel", bundle.getString("alerter.notify.target"));
-        context.setVariable("target", alert.getTarget());
+        String monitorId=alert.getTags().get("monitorId");
```
Comment (gcdd1993) [inline manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/AbstractAlertNotifyHandlerImpl.java:78]: 存在空指针风险

**Review by gcdd1993 — CHANGES_REQUESTED** [review]

**Review by gcdd1993 — APPROVED** [review]

**Review by gcdd1993 — APPROVED** [review]

**Comment by tomsun28** [comment]
hi pls fix these confilcts
```
manager/pom.xml
web-app/src/assets/i18n/zh-CN.json
```

**Comment by tomsun28** [comment]
- 启动修改 hzb_notice_rule 表新增 template_id 字段报错不能为null, 这里是因为之前的老数据没有此字段导致，新增 `alter table hzb_notice_rule add column template_id bigint not null` 会导致之前的老数据异常
```
2023-10-11 11:14:23 [main] WARN  org.hibernate.tool.schema.internal.ExceptionHandlerLoggedImpl - GenerationTarget encountered exception accepting command : Error executing DDL "alter table hzb_notice_rule add column template_id bigint not null" via JDBC Statement
org.hibernate.tool.schema.spi.CommandAcceptanceException: Error executing DDL "alter table hzb_notice_rule add column template_id bigint not null" via JDBC Statement
```
建议不设置 `template_id` 约束null，然后考虑老数据问题，当template_id为null时应该有个对应的默认模版

- 启动时报错 'classpath:sql/data.sql' 找不到资源，那是因为没有在manager模块的pom.xml里面配置资源路径，导致编译class时没有主动编译此资源到类路径下面去 
`Caused by: java.lang.IllegalStateException: No data scripts found at location 'classpath:sql/data.sql'`
需要把 `<include>sql/**</include>` 添加到 manager pom.xml `build.resources.resource.includes` 下

**Comment by tomsun28** [comment]
要不要把我们内置的默认消息模版像之前一样以文件的方式存储，用户自定义消息模版用现在的数据库方式存储。系统启动时从文件中加载内置模版到内存中，当 `template_id` 为null找不到时就使用我们内置的消息模版。页面配置用户不选择的情况下默认就使用内置消息模版。
这样能避免 template_id 为null老数据问题和data.sql重复初始化等问题，无需初始化数据库里面的消息模版数据。

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
Should we store our built-in default message templates in files as before, and user-defined message templates in the current database method? When the system starts, the built-in template is loaded from the file into the memory. When `template_id` is null and cannot be found, our built-in message template is used. If the page configuration user does not select it, the built-in message template will be used by default.
This can avoid the problem of old data with template_id being null and repeated initialization of data.sql. There is no need to initialize the message template data in the database.

**Comment by Eden4701** [comment]
> 要不要把我们内置的默认消息模版像之前一样以文件的方式存储，用户自定义消息模版用现在的数据库方式存储。系统启动时从文件中加载内置模版到内存中，当 `template_id` 为null找不到时就使用我们内置的消息模版。页面配置用户不选择的情况下默认就使用内置消息模版。 这样能避免 template_id 为null老数据问题和data.sql重复初始化等问题，无需初始化数据库里面的消息模版数据。

之前阿晨老师说不想维护文件配置，所以全部入库了，但我没有考虑到老数据的问题，这个方案也可以

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
> Should we store our built-in default message templates in files as before, and user-defined message templates in the current database method. When the system starts, the built-in template is loaded from the file into the memory. When `template_id` is null and cannot be found, our built-in message template is used. If the page configuration user does not select it, the built-in message template will be used by default. This can avoid the problem of old data with template_id being null and repeated initialization of data.sql. There is no need to initialize the message template data in the database.

Previously, Teacher Achen said that he did not want to maintain file configurations, so he put them all into the database, but I did not consider the problem of old data. This solution can also be used

**Comment by gcdd1993** [comment]
@tomsun28 @Eden4701 可以换成用文件存储，这样就不用初始化数据和是否为默认模板这个字段了；

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
@tomsun28 @Eden4701 It can be changed to file storage, so that there is no need to initialize the data and the fields of whether it is the default template;

**Comment by tomsun28** [comment]
> @tomsun28 @Eden4701 可以换成用文件存储，这样就不用初始化数据和是否为默认模板这个字段了；但是缺点是需要额外维护一下从文件获取模板的代码，而且，如果要在页面显示默认模板的话，还需要额外编写代码（文件 + 数据库的分页感觉有点难做）

是的，如果这样做的话看能不能应用层的代码尽量不改动，主要是启动时加载内置的文件模版，还有数据层操作，因为内置模版只查询不修改，数据层分页改造的时候我们固定把内置模版排序到最前面，然后计算剩余pagesize 和pagenum查询数据库数据后合并 用 `new PageImpl<>(list)`来返回数据

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
> @tomsun28 @Eden4701 It can be replaced by file storage, so that there is no need to initialize the data and whether it is the default template field; but the disadvantage is that it requires additional maintenance of the code to obtain the template from the file, and if you want to display the default template on the page , additional code needs to be written (file + database paging feels a bit difficult to do)

Yes, if you do this, see if you can keep the code of the application layer as unchanged as possible. The main thing is to load the built-in file template at startup, and to operate the data layer. Because the built-in template only queries and does not modify, we fixed it during the data layer paging transformation. Sort the built-in templates to the front, then calculate the remaining pagesize and pagenum to query the database data and then merge them using `new PageImpl<>(list)` to return the data.

**Comment by Eden4701** [comment]
> > @tomsun28 @Eden4701 可以换成用文件存储，这样就不用初始化数据和是否为默认模板这个字段了；但是缺点是需要额外维护一下从文件获取模板的代码，而且，如果要在页面显示默认模板的话，还需要额外编写代码（文件 + 数据库的分页感觉有点难做）
> 
> 是的，如果这样做的话看能不能应用层的代码尽量不改动，主要是启动时加载内置的文件模版，还有数据层操作，因为内置模版只查询不修改，数据层分页改造的时候我们固定把内置模版排序到最前面，然后计算剩余pagesize 和pagenum查询数据库数据后合并 用 `new PageImpl<>(list)`来返回数据

新的代码已提交，预设模板文件存储，自定义模板在数据库，设置通知规则时模板是可选项，用户不选择模板时默认使用预设模板

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
> > @tomsun28 @Eden4701 You can switch to file storage, so you don’t need to initialize the data and the fields of whether it is the default template; but the disadvantage is that you need to maintain the code to get the template from the file, and if you want to display the default template on the page If so, additional code needs to be written (file + database paging seems a bit difficult to do)
>
> Yes, if you do this, see if you can keep the application layer code as unchanged as possible. The main thing is to load the built-in file template at startup and data layer operations, because the built-in template only queries and does not modify. When we transform the data layer paging Fixed sorting the built-in templates to the front, then calculating the remaining pagesize and pagenum, querying the database data, and then merging and using `new PageImpl<>(list)` to return the data.

The new code has been submitted, the preset template file is stored, and the custom template is in the database. The template is optional when setting notification rules. When the user does not select a template, the preset template is used by default.

File: manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/AbstractAlertNotifyHandlerImpl.java [hunk 6]
```diff
@@ -17,78 +17,97 @@
...
      *
      * @return Thymeleaf模板名称
      */
     protected abstract String templateName();
-    
-    @EventListener(SystemConfigChangeEvent.class)
-    public void onEvent(SystemConfigChangeEvent event) {
-        log.info("{} receive system config change event: {}.", this.getClass().getName(), event.getSource());
-        this.bundle = ResourceBundleUtil.getBundle("alerter");
-    }
```
Comment (tomsun28) [inline manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/AbstractAlertNotifyHandlerImpl.java:93]: hi 这里需要保留 当用户语言变更时 模版语言监听变更

File: manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/EmailAlertNotifyHandlerImpl.java [hunk 7]
```diff
@@ -17,113 +17,52 @@
...
-                    sender.setHost(host);
-                    sender.setPort(port);
-                    sender.setUsername(username);
-                    sender.setPassword(password);
-                    Properties props = sender.getJavaMailProperties();
-                    props.put("mail.smtp.ssl.enable", sslEnable);
-                }
-            } catch (Exception e) {
-                log.error("Type not found {}",e.getMessage());
-            }
```
Comment (tomsun28) [inline manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/EmailAlertNotifyHandlerImpl.java:116]: hi 数据库获取邮件服务器配置需要保留，我们现在支持用户在页面配置邮件服务器 或者 在 application.yml配置邮件服务器。优先级 页面数据库配置 高于 application.yml 文件配置。

File: manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/EmailAlertNotifyHandlerImpl.java [hunk 8]
```diff
@@ -136,10 +75,4 @@ public void send(NoticeReceiver receiver, Alert alert) throws AlertNoticeExcepti
     public byte type() {
         return 1;
     }
-    
-    @EventListener(SystemConfigChangeEvent.class)
-    public void onEvent(SystemConfigChangeEvent event) {
-        log.info("{} receive system config change event: {}.", this.getClass().getName(), event.getSource());
-        this.bundle = ResourceBundleUtil.getBundle("alerter");
-    }
```
Comment (tomsun28) [inline manager/src/main/java/org/dromara/hertzbeat/manager/component/alerter/impl/EmailAlertNotifyHandlerImpl.java:144]: hi 同上上，这里监听语言变更也需要保留

File: manager/src/main/java/org/dromara/hertzbeat/manager/service/impl/MailServiceImpl.java [hunk 9]
```diff
@@ -65,31 +67,39 @@ public String buildAlertHtmlTemplate(final Alert alert) {
...
         SimpleDateFormat simpleDateFormat = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss");
-        String alarmTime = simpleDateFormat.format(new Date(alert.getLastAlarmTime()));
-        context.setVariable("lastTriggerTime", alarmTime);
-        return templateEngine.process("mailAlarm", context);
-    }
-    
-    @EventListener(SystemConfigChangeEvent.class)
-    public void onEvent(SystemConfigChangeEvent event) {
-        log.info("{} receive system config change event: {}.", this.getClass().getName(), event.getSource());
-        this.bundle = ResourceBundleUtil.getBundle("alerter");
```
Comment (tomsun28) [inline manager/src/main/java/org/dromara/hertzbeat/manager/service/impl/MailServiceImpl.java:93]: hi 同上，这里监听语言变更也需要保留

**Review by tomsun28 — APPROVED** [review]
👍👍👍LGTM

**Comment by tomsun28** [comment]
赞👍 完成度和质量都很高，这个PR合并了哦。
后续UI优化建议供参考：
1. [通知模板] 菜单放到第三位即 [告警通知策略] 后面，因为通知模版在默认情况下用户不修改也能使用，对于多数没有修改模版需求的用户它只需前面两步即可。
2. [告警通知策略]新增配置时建议[通知模版框]根据选中的接收人的通知类型自动渲染填充系统内置默认模版
3. 编辑通知模版UI建议有展示当前所支持的模版环境变量，参考华为云消息通知模版
image

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
Like 👍 The completion and quality are very high, this PR has been merged.
Follow-up UI optimization suggestions are for reference:
1. The [Notification Template] menu is placed in the third position, that is, after [Alarm Notification Policy], because the notification template can be used by users without modification by default. For most users who do not need to modify the template, it only requires the first two steps. .
2. [Alarm Notification Policy] When adding new configuration, it is recommended that [Notification Template Box] automatically render and fill the system's built-in default template according to the notification type of the selected recipient.
3. It is recommended to display the currently supported template environment variables when editing the notification template UI. Please refer to Huawei Cloud Message Notification Template.
image

**Comment by tomsun28** [comment]
@all-contributors please add @Eden4701  for code

**Comment by Eden4701** [comment]
> 赞👍 完成度和质量都很高，这个PR合并了哦。 后续UI优化建议供参考：
> 
> 1. [通知模板] 菜单放到第三位即 [告警通知策略] 后面，因为通知模版在默认情况下用户不修改也能使用，对于多数没有修改模版需求的用户它只需前面两步即可。
> 2. [告警通知策略]新增配置时建议[通知模版框]根据选中的接收人的通知类型自动渲染填充系统内置默认模版
> 3. 编辑通知模版UI建议有展示当前所支持的模版环境变量，参考华为云消息通知模版
> 
> image

收到！

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
> Like 👍 The completion and quality are very high, this PR has been merged. Follow-up UI optimization suggestions are for reference:
>
> 1. The [Notification Template] menu is placed in the third position, that is, after [Alarm Notification Policy], because the notification template can be used by users without modification by default. For most users who do not need to modify the template, it only requires the first two steps. Can.
> 2. [Alarm Notification Policy] When adding new configurations, it is recommended that the [Notification Template Box] automatically render and fill the system's built-in default template according to the notification type of the selected recipient.
> 3. It is recommended to display the currently supported template environment variables when editing the notification template UI. Please refer to Huawei Cloud Message Notification Template.
>
> image

receive!

**Comment by tomsun28** [comment]
@Eden4701 hi 打扰了 今天使用中有个建议反馈下 我们现在对默认模版每次是发告警时去打开文件，读取文件内容，关闭文件这样，在大量告警消息时会很耗性能。建议参考 org.dromara.hertzbeat.manager.service.impl.AppServiceImpl#run 在程序启动时统一读取文件加载到内存中，发告警消息时若使用内置默认模版，就到内存中取模版内容即可。

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
@Eden4701 hi I’m sorry to bother you. I have a suggestion during use today. Please give me feedback. We now use the default template to open the file, read the file content, and close the file every time an alarm is sent. This will consume a lot of performance when there are a large number of alarm messages. It is recommended to refer to org.dromara.hertzbeat.manager.service.impl.AppServiceImpl#run to uniformly read the file and load it into the memory when the program starts. If you use the built-in default template when sending an alarm message, just get the template content from the memory.

**Comment by Eden4701** [comment]
> @Eden4701 hi 打扰了 今天使用中有个建议反馈下 我们现在对默认模版每次是发告警时去打开文件，读取文件内容，关闭文件这样，在大量告警消息时会很耗性能。建议参考 org.dromara.hertzbeat.manager.service.impl.AppServiceImpl#run 在程序启动时统一读取文件加载到内存中，发告警消息时若使用内置默认模版，就到内存中取模版内容即可。

ok

**Comment by hertzbeat** [comment]
> Bot detected the issue body's language is not English, translate it automatically. 👯👭🏻🧑‍🤝‍🧑👫🧑🏿‍🤝‍🧑🏻👩🏾‍🤝‍👨🏿👬🏿         
----    
> @Eden4701 hi Sorry to bother you. I have a suggestion and feedback during use today. We now use the default template to open the file, read the file content, and close the file every time an alarm is sent. This will consume a lot of performance when there are a large number of alarm messages. It is recommended to refer to org.dromara.hertzbeat.manager.service.impl.AppServiceImpl#run to uniformly read the file and load it into the memory when the program starts. If you use the built-in default template when sending an alarm message, just get the template content from the memory.

OK

**Pull request merged**